src/main/java/be/jidoka/jdk/keycloak/admin/service/KeycloakUserAdminService.java

```diff
@@ -78,15 +85,41 @@ public Page<User> searchUsers(SearchUserRequest request) {
 	}
 
 	public Set<User> searchUsersByClientRole(SearchUserByClientRoleRequest request) {
-		return clientsResource.get(request.getClientId()).roles()
-				.get(request.getRoleName())
-				.getRoleUserMembers()
-				.stream()
+		return Stream.concat(
+						getUserRepresentationsWithImplicitRole(request),
+						getUserRepresentationsWithExplicitRole(request)
+				)
 				.map(user -> enhanceWithClientRoles(user, request.getClientId()))
 				.map(user -> new User(user, request.getClientId()))
 				.collect(Collectors.toSet());
 	}
 
+	private Stream<UserRepresentation> getUserRepresentationsWithExplicitRole(SearchUserByClientRoleRequest request) {
+		RolesResource rolesResource = clientsResource.get(request.getClientId()).roles();
+
+		if (rolesResource.list().stream().anyMatch(roleRepresentation -> roleRepresentation.getName().equals(request.getRoleName()))) {
+			RoleResource roleResource = rolesResource.get(request.getRoleName());
+
+			return roleResource.getRoleUserMembers().stream();
+		}
+
+		return Stream.of();
+	}
+
+	private Stream<UserRepresentation> getUserRepresentationsWithImplicitRole(SearchUserByClientRoleRequest request) {
+		List<GroupRepresentation> groupsWithGivenRole = groupsResource.groups().stream()
+				.map(groupRepresentation -> enhanceWithClientRoles(groupRepresentation, request.getClientId()))
+				.filter(groupRepresentation -> groupRepresentation.getClientRoles().get(request.getClientId()).contains(request.getRoleName()))
+				.toList();
+
+		return groupsWithGivenRole.stream()
```

Comment:
In theory not needed to use a variable named `groupsWithGivenRole`, if we immediately gonna stream on it again.